Mathlib/Algebra/Polynomial/Factors.lean

```diff
@@ -0,0 +1,284 @@
+      or_false, false_or, ← Multiset.prod_toList, List.prod_isUnit_iff] at h
+    have : m = 0 := by simpa [not_isUnit_X_add_C, ← Multiset.eq_zero_iff_forall_notMem] using h.1.2
+    grw [hm, this, natDegree_mul_le]
+    simp
+
+end CommSemiring
+
+section Ring
+
+variable [Ring R]
+
+@[simp, aesop safe apply]
+theorem factors_X_sub_C (a : R) : Factors (X - C a) := by
+  simpa using factors_X_add_C (-a)
+
+@[aesop safe apply]
+protected theorem Factors.neg {f : R[X]} (hf : Factors f) : Factors (-f) := by
+  rw [← neg_one_mul, ← C_1, ← C_neg]
+  exact hf.C_mul (-1)
+
+@[simp]
+theorem factors_neg_iff {f : R[X]} : Factors (-f) ↔ Factors f :=
+  ⟨fun hf ↦ neg_neg f ▸ hf.neg, .neg⟩
+
+end Ring
+
+section CommRing
+
+variable [CommRing R]
+
+theorem factors_iff_exists_multiset {f : R[X]} :
+    Factors f ↔ ∃ m : Multiset R, f = C f.leadingCoeff * (m.map (X - C ·)).prod := by
+  refine factors_iff_exists_multiset'.trans ⟨?_, ?_⟩ <;>
+    rintro ⟨m, hm⟩ <;> exact ⟨m.map (- ·), by simpa⟩
+
+theorem exists_root_of_factors {f : R[X]} (hf : Factors f) (hf0 : degree f ≠ 0) :
+    ∃ a, eval a f = 0 := by
+  obtain ⟨m, hm⟩ := factors_iff_exists_multiset.mp hf
+  by_cases hf₀ : f.leadingCoeff = 0
+  · simp [leadingCoeff_eq_zero.mp hf₀]
+  obtain rfl | ⟨a, ha⟩ := m.empty_or_exists_mem
+  · rw [hm, Multiset.map_zero, Multiset.prod_zero, mul_one, degree_C hf₀] at hf0
+    contradiction
+  obtain ⟨m, rfl⟩ := Multiset.exists_cons_of_mem ha
+  exact ⟨a, hm ▸ by simp⟩
+
+variable [IsDomain R]
+
+theorem Factors.eq_prod_roots {f : R[X]} (hf : Factors f) :
+    f = C f.leadingCoeff * (f.roots.map (X - C ·)).prod := by
+  by_cases hf0 : f.leadingCoeff = 0
+  · simp [leadingCoeff_eq_zero.mp hf0]
+  · obtain ⟨m, hm⟩ := factors_iff_exists_multiset.mp hf
+    suffices hf : f.roots = m by rwa [hf]
+    rw [hm, roots_C_mul _ hf0, roots_multiset_prod_X_sub_C]
+
+theorem Factors.natDegree_eq_card_roots {f : R[X]} (hf : Factors f) :
+    f.natDegree = f.roots.card := by
+  by_cases hf0 : f.leadingCoeff = 0
+  · simp [leadingCoeff_eq_zero.mp hf0]
+  · conv_lhs => rw [hf.eq_prod_roots, natDegree_C_mul hf0, natDegree_multiset_prod_X_sub_C_eq_card]
+
+theorem roots_ne_zero_of_factors {f : R[X]} (hf : Factors f) (hf0 : natDegree f ≠ 0) :
+    f.roots ≠ 0 := by
+  obtain ⟨a, ha⟩ := exists_root_of_factors hf (degree_ne_of_natDegree_ne hf0)
+  exact mt (· ▸ (mem_roots (by aesop)).mpr ha) (Multiset.notMem_zero a)
+
+theorem factors_X_sub_C_mul_iff {f : R[X]} {a : R} : Factors ((X - C a) * f) ↔ Factors f := by
+  refine ⟨fun hf ↦ ?_, ((factors_X_sub_C _).mul ·)⟩
+  by_cases hf₀ : f = 0
+  · aesop
+  have := hf.eq_prod_roots
+  rw [leadingCoeff_mul, leadingCoeff_X_sub_C, one_mul,
+    roots_mul (mul_ne_zero (X_sub_C_ne_zero _) hf₀), roots_X_sub_C,
+    Multiset.singleton_add, Multiset.map_cons, Multiset.prod_cons, mul_left_comm] at this
+  rw [mul_left_cancel₀ (X_sub_C_ne_zero _) this]
+  aesop
+
+theorem factors_mul_iff {f g : R[X]} (hfg : f * g ≠ 0) :
```

Comment:
Since we are already in a domain I think it would be more useful to state the condition as `(hf : f ≠ 0) (hg : g ≠ 0)`.